circuitbreaker.py

```diff
@@ -15,28 +15,81 @@
 except ImportError:
     from monotonic import monotonic
 
+# Python2 vs Python3 strings
+try:
+    basestring
+    STRING_TYPES = (basestring, unicode)
+except NameError:
+    STRING_TYPES = (bytes, str)
+
+
 STATE_CLOSED = 'closed'
 STATE_OPEN = 'open'
 STATE_HALF_OPEN = 'half_open'
 
+def in_exception_list(*exc_types):
+    """Build a predicate function that checks if an exception is a subtype from a list"""
+    def matches_types(thrown_type, thrown_value):
+        return issubclass(thrown_type, exc_types)
+    return matches_types
 
 class CircuitBreaker(object):
     FAILURE_THRESHOLD = 5
     RECOVERY_TIMEOUT = 30
-    EXPECTED_EXCEPTION = Exception
     FALLBACK_FUNCTION = None
 
     def __init__(self,
                  failure_threshold=None,
                  recovery_timeout=None,
                  expected_exception=None,
                  name=None,
-                 fallback_function=None):
+                 fallback_function=None
+                 ):
+        """
+        Construct a circuit breaker.
+
+        :param failure_threshold: break open after this many failures
+        :param recovery_timout: close after this many seconds
+        :param expected_exception: either an type of Exception, iterable of Exception types, or a predicate function.
+          If an Exception type or iterable of Exception types, a failure will be triggered when a thrown
+          exception matches a type.
+
+          If a predicate function, it will be called with the exception type and the exception value 
+          when an exception is thrown. It should have the signature (Type[Exception], Exception) -> bool.
+          Return value True indicates a failure in the underlying function.
+
+        :param name: name for this circuitbreaker
+        :param fallback_function: called when the circuit is opened
+
+           :return: Circuitbreaker instance
+           :rtype: Circuitbreaker
+        """
         self._last_failure = None
         self._failure_count = 0
         self._failure_threshold = failure_threshold or self.FAILURE_THRESHOLD
         self._recovery_timeout = recovery_timeout or self.RECOVERY_TIMEOUT
-        self._expected_exception = expected_exception or self.EXPECTED_EXCEPTION
+
+        # default to plain old Exception
+        expected_exception = expected_exception or Exception
+
+        # auto-construct a failure predicate, depending on the type of the 'expected_exception' param
+        if isclass(expected_exception) and issubclass(expected_exception, Exception):
```

Comment:
I would prefer factoring this out into a separate function, to reduce the complexity of the constructor again. What do you think?

Reply:
Separate `build_predicate_function` created.